### ISSUE TITLE
[RFC/Discussions] AutoMapper v8

### ISSUE THREAD

**Issue by nartc** [nartc/mapper#423]
As the title said, this issue acts as a discussion thread for V8. If there's no discussion going on, I'll still be using it to keep you guys updated.

### Related Issues
- #319
- #333, #358 

### Proxy-based

It will be a completely rewrite using `Proxy` to enable lazy-initialization as much as possible.

```ts
// before
function createMapper() {

   // already initialized even if addTypeConverter is never used
   const typeConverters = new Map();    

   return {
      addTypeConverter() {/*..*/}
   }
}

// after
function createMapper() {
   // only declare the variable
   let typeConverters: Map;

   return new Proxy({}, {
      get(target, prop, receiver) {
         if (prop === 'typeConverters') {
            if (!typeConverters) {
               // will only get initialized when it is first asked for
               typeConveters = new Map();
            }
            return typeConverters;
         }
      }
   })
}
```

### Breaking Changes

The API of `AutoMapper` will change drastically to be:
- As tree-shakable as possible
- Lazy-initialized when needed

#### Functional approach

V8 will attempt to ditch the Fluent API approach (I am sorry .NET folks 😅, but I think it will be worth it) to go for a more functional approach. 

Let's recap on how things work in `AutoMapper` today

```ts
// create the mapper
const mapper = createMapper();

// Fluent API
mapper
   // (1)
   .addTypeConverter()
   // (2)
   .createMap(Source, Destination, someOptions)
   // (3)
   .forMember()
   .beforeMap()
   .afterMap();
```

- (1): If you need to create some `TypeConverter`, you need to do it on the Mapper level. This means that `Mapping` for `SourceOne` and `DestinationOne` will share the `Type Converters` with `SourceTwo` and `DestinationTwo`. That might not ALWAYS be the case.
- (2): You always need to create the Mapping first. Calling `createMap()` always initializes the Mapping for `Source` and `Destination` eagerly. This makes it IMPOSSIBLE to have Custom Constructor. This also introduces unnecessary code duplication. When a Mapping is initialized, it will assign a DEFAULT mapping instruction (`MapInitialize`) to ALL properties that are subject to `AutoMap`. Then, the mapping instruction of any given property CAN BE overridden by `forMember()`.
- (3): These are chain-able off of `createMap()`. The order can get weird because `beforeMap()` and `afterMap()` need to be chain-able but they can be literally anywhere in the chain. It looks confusing when chaining `afterMap().forMember().beforeMap()`.

And to top it off, all these Fluent API methods aren't tree-shakeable because they are always present on the `Mapper`.

Let's look at how AutoMapper V8 might look like (this is still WIP so things might change):

```ts
// create the mapper
const mapper = createMapper();

// createMap is now a function that accepts the Mapper instead
// If you never need createMap (by mistake, you create the Mapper but don't need to use it)
// you don't pay for createMap()
createMap(mapper, Source, Destination)
```

How will `forMember`, `afterMap`, `beforeMap`, etc... look like? They are all functions

```ts
createMap(
   mapper, 
   Source, 
   Destination,
   withOptions(), // provide the options for this Mapping
   forMember(), // configure mapping instruction for a Member
   typeConverter(), // configure a Type Converter for this Mapping (remember mapping level)
   beforeMap(), // add a beforeMap(), now you can add MULTIPLE callbacks
   afterMap(), // add an afterMap(), now you can add MULTIPLE callbacks
   // now the juicy part
   constructUsing(), // teach this Mapping how to initialize a Model (Source or Destination)
)
```

Since they're all separate functions, you only pay for what you actually use. And best of all, they allow for `createMap()` to be lazy. Internally, each of these functions returns a `MappingConfig` object which:
- Based on the `MappingConfig` type, a Mapping will be initialized with the FINAL configurations
- `forMember()` will be used to INITIALIZE the Mapping instead of overriding. No duplication required
- `typeConverter()` can be applied to a Mapping level instead of just Mapper level
- `constructUsing()` can be used to create Custom Constructor. Since a Mapping can be created AFTER all the configurations, the Custom Constructor can actually participate in the initialization process.

#### Map

Currently, there are four APIs for mapping: `map`, `mapAsync`, `mapArray`, and `mapArrayAsync`. In AutoMapper v8, there will be four more: `mutate`, `mutateAsync`, `mutateArray`, and `mutateArrayAsync`. This is to separate the mutation apart from the original `map`

#### Async

TBD but expect a better story about Async mapping.

#### Plugins

https://github.com/nartc/mapper/issues/423#issuecomment-1061374723

### Documentations

Documentations will be rewritten and will have runable examples for many different cases.

### Migrations

I'll try my best to provide auto migrations on trivial parts. But please rest assured that I will provide, as detail as possible, migration guides.

That's all I have for now. Any thoughts/contributions are definitely appreciated.

**Comment by nartc** [comment]
## Plugins

### Overall

I'll be renaming `Plugins` into `Strategies`. Right now, a plugin isn't really a "plugin" per se. They are more like a "Mapping Strategy", you specify a strategy for a certain Mapper. 

The API of Strategy is **SIGNIFICANTLY** less complicated than the old Plugin's API, with better extensibility as well

```ts
// before
export function classes(errorHandler: ErrorHandler) {
 /* ... */
}

// after
export function classes(applyMetadata?, destinationConstructor?) {
   return (mapper) => {}
}
```

**Comment by nartc** [comment]
## Map

`map()` API will change from: `map(sourceObject, Destination, Source)` to `map(sourceObject, Source, Destination)`. This is to address the confusion, especially for `pojos`.

```ts
// before
map<Foo, FooDto>(foo, 'FooDto', 'Foo'); // weird order

// after
map<Foo, FooDto>(foo, 'Foo', 'FooDto'); // matching order
```

### Extra Arguments

`extraArguments` has been renamed to `extraArgs` and it is a Function instead of just a `Record<string, unknown>`. 

```ts
// before
map(foo, FooDto, Foo, { extraArguments: {some: 'extra'} });

// after
map(foo, Foo, FooDto, { extraArgs: (mapping, destination) => ({ some: 'extra' }) })
```

`extraArgs` function is invoked with the `Mapping` and the current `Destination` object which give the consumers all the information they need. `Mapper` is already available (because you would need to use: `mapper.map()`), the `Source` object is already available (because you would need to invoke `mapper.map()` with the `sourceObject`). `(mapping, destination)` are two things that you don't have access to "easily".

### destinationConstructor

In addition to `{ beforeMap, afterMap, extraArgs}`, `MapOptions` now includes `destinationConstructor` which the consumers can use to tell AutoMapper how to construct the `Destination` for a `map()` operation.

```ts
map(foo, Foo, FooDto, { destinationConstructor: () => new FooDto(arg1, arg2, arg3) })
```

**Comment by AliYusuf95** [comment]
Hi @nartc, I'm very excited about the new features in V8.

I have two more enhancements:
- Strict array types
- Support arrays in `addTypeConverter`
- `MappingConfig` profile level


### Strict array types:
I suggest that we should explicitly define if the property is an array or not instead of relying on the mapper to finger out it is an array. I think with this approche we'll avoid unexpected behavior of assuming and enhance the error messages.

```ts
class Foo {
    @AutoMap({ typeFn: () => [String] })
    fooArr!: String[];
}

interface IFoo {
    fooArr!: String[];
}

createMetadataMap<IFoo>('IFoo', {
    fooArr: [String],
});

```

Unexpected behavior example:
```ts
class OptionObj {
    id!: string;
}

class Foo {
    @AutoMap({ typeFn: () => OptionObj })
    fooArr!: OptionObj[];
}

class FooDto {
    @AutoMap()
    fooArr!: string;
}

mapper.addTypeConverter(OptionObj, String, (src) => src.id);

mapper.createMap(Foo, FooDto);

const opt = new OptionObj();
opt.id = '1';

const foo = new Foo();
foo.fooArr = [opt, opt, opt];

console.log(mapper.map(foo, FooDto, Foo).fooArr); // prints undefined
// should throw error
// Mapping is not found for [OptionObj] and String

// and with this converter
mapper.addTypeConverter(OptionObj, String, (src: any) => src.map((i: any) => i.id));
console.log(mapper.map(foo, FooDto, Foo).fooArr); // prints [ '1', '1', '1' ]
```

### Support arrays in `addTypeConverter`:
Currently, it works only with primitive types and classes but I think it would be more useful if it can support arrays as well.

For example:
```ts
class OptionObj {
    id!: string;
}

class Foo {
    @AutoMap({ typeFn: () => [OptionObj] })
    fooArr!: OptionObj[];
}

class FooDto {
    @AutoMap()
    fooArr!: string;
}

mapper.addTypeConverter([OptionObj], [String], (src) => src.map(ob => ob.id));

mapper.createMap(Foo, FooDto);

const opt = new OptionObj();
opt.id = '1';

const foo = new Foo();
foo.fooArr = [opt, opt, opt];

const dto = mapper.map(foo, FooDto, Foo);

console.log(dto.fooArr); // should print [ '1', '1', '1' ]

```

### `MappingConfig` profile level:

It's nice how the new functional approach works, I think with this approach it will be easier to utilize the profile and use it to pass aditinal configrations e.g. `typeConverter()` that will affect only the profile context mappings.

```ts

function fooProfile(mapper) {
    //  mapper object with profile context
    createMap(mapper, Source, Destination);
}

addProfile(
    mapper,
    fooProfile,
    typeConverter(), // this converter will be added to all foo profile mappings
)

```

**Comment by nartc** [comment]
@AliYusuf95 Thank you for the suggestions (and the kind words).

<img width="574" alt="image" src="https://user-images.githubusercontent.com/25516557/157463642-8baaea35-314d-4d16-b02e-be72d30c6d9e.png">
Yeah, I think we're on the same page about String Array type. No lie, I have just decided to make the Array type explicit last night. I also added an overload to make `AutoMap()` easier to use. `typeFn` sounds weird so that got renamed as well to just `type` (I've already have a plan to create auto migrations for these).

Additionally, now that AutoMapper won't try to `instantiate` the model anymore, additional metadata like `Job` here in the example will need to be handled separately with `AutoMapStandalone()`

<img width="343" alt="image" src="https://user-images.githubusercontent.com/25516557/157464349-8f65b968-808f-45ae-bfe9-1b930f7318ef.png">

I like `MappingConfig` profile level. I think that definitely is possible with the new approach.

However, my roadblock right now is some typing issue with this new functional approach. A couple of `mapMemberFn` (`convertUsing`, `ignore`, etc...) typings get weird. I'm also considering dropping `mapDefer` completely.

**Comment by nartc** [comment]
@AliYusuf95 if you have some free time, can you please take a look at `address.profile.ts` on this branch https://github.com/nartc/mapper/tree/rewrite ? I somewhat know why the typing issue is happening but not sure how to fix it :(

**Comment by AliYusuf95** [comment]
@nartc 
I think typescript somehow joins the second parameter type of the functions which leads to a wrong destination type.
The issue is caused by in `MapFromReturn` second parameter `destination?: TDestination`.
Using a different generic type for it will solve the issue.

```ts
export type MapFromReturn<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>,
    TDestinationParam = TDestination
> = [
    TransformationType.MapFrom,
    (source: TSource, destination?: TDestinationParam) => TSelectorReturn
];
```

**Comment by nartc** [comment]
@AliYusuf95 I fixed it.The `destination` parameter isn't used anyway so I just removed that one. Do you think `mapDefer` is a use-case worth keeping? I've never used it personally.

**Comment by micalevisk** [comment]
@nartc I've been using `mapDefer` in one profile of mine

![image](https://user-images.githubusercontent.com/13461315/157713682-28c9b329-e6c3-4cf3-b99b-e5f26bde5ac1.png)

you see that I didn't use the `source` parameter so I might be able to get ride of the `mapDefer`

I don't recall why I need to use it there, tbh.

**Comment by nartc** [comment]
@micalevisk right. In this case, I think you can just use `mapWith` directly. I'm thinking of dropping `mapDefer` because `MapDeferReturn` complicates the type a lot :(

**Comment by micalevisk** [comment]
I just realise that I have to use `mapDefer` due to this #306 btw

**Comment by nartc** [comment]
@AliYusuf95 `MappingProfile` is pretty pure at the moment. 

```ts
const fooProfile: MappingProfile = mapper => {
   createMap(mapper, Foo, FooDto);
}
```
then `addProfile`

```ts
function addProfile(mapper, profile: MappingProfile) {
   profile(mapper);
}
```
I mean, now that I write it out this way, `addProfile` is **unneccessary** because you can just:

```ts
fooProfile(mapper);
// instead of: addProfile(mapper, fooProfile);
```

What makes sense right now is to adjust `MappingProfile` type to also accept a list of `MappingConfiguration` then the consumers can spread those into their `createMap()`

```ts
fooProfile(
   mapper,
   typeConverter(...),
   typeConverter(...)
)

// 

const fooProfile: MappingProfile = (mapper, ...mappingConfigurations) => {
   createMap(mapper, ..., ...mappingConfigurations);
   createMap(mapper, ..., ...mappingConfigurations);
}
```

What are your thoughts? Any API proposal?

**Comment by AliYusuf95** [comment]
@nartc with this approach as you said `addProfile` will be **unnecessary** unless it's bring some added value.
My first thought was to make the profile have some context to make it a bit easier to pass configs without explicitly writing them on each `createMap` call.

It could be something like this
```ts
function addProfile(mapper, profile: MappingProfile, ...mappingConfigurations) {
  // start context; where the mapper will apply the configs in next `createMap` calls
   profile(mapper);
  // end of the context
}

const fooProfile: MappingProfile = mapper => {
    // the mapper will apply the typeConverters implecitly
   createMap(mapper, Foo, FooDto);
}

addProfile(
   mapper,
   fooProfile,
   typeConverter(...),
   typeConverter(...)
)
```

It sounds like an inheritance where we can allow the context to be extended in nested profiles

```ts
const fooProfile: MappingProfile = mapper => {
   // mapper will have typeConverter A
   createMap(mapper, Foo, FooDto);
   addProfile(
      mapper,
      barProfile,
      typeConverter(...) // B
   )
}

const barProfile: MappingProfile = mapper => {
   // mapper will have typeConverter A and B
   createMap(mapper, Bar, BarDto);
}

addProfile(
   mapper,
   fooProfile,
   typeConverter(...) // A
)

```
---
I think the most case where `mapDefer` will be used is for discriminators along with `mapWith`. Actually, I have used it a couple of times in my last project. while the same result can be achieved with conditions but I think `mapDefer` is more readable.

```ts
class Foo {
    @AutoMap()
    event!: Event;
}

class FooDTO {
    @AutoMap()
    event!: EventDTO;
}

class Event {
    @AutoMap()
    id!: string;
    @AutoMap()
    type!: 'click' | 'other';
}

class EventDTO {
    @AutoMap()
    id!: string;
    @AutoMap()
    type!: 'click' | 'other';
}

class ClickEvent extends Event {
    @AutoMap()
    x!: number;
    @AutoMap()
    y!: number;
}

class ClickEventDTO extends EventDTO {
    @AutoMap()
    x!: number;
    @AutoMap()
    y!: number;
}

class OtherEvent extends Event {
    @AutoMap()
    data!: string;
}

class OtherEventDTO extends EventDTO {
    @AutoMap()
    data!: string;
}

Mapper.createMap(Foo, FooDTO).forMember(
    (d) => d.event,
    mapDefer((source) => {
        if (source.event.type === 'click') {
            return mapWith(ClickEventDTO, ClickEvent, source.event);
        }
        return mapWith(OtherEventDTO, OtherEvent, source.event);
    }),
);
Mapper.createMap(Event, EventDTO);
Mapper.createMap(ClickEvent, ClickEventDTO, {
    extends: [Mapper.getMapping(Event, EventDTO)],
});
Mapper.createMap(OtherEvent, OtherEventDTO, {
    extends: [Mapper.getMapping(Event, EventDTO)],
});

```

**Comment by nartc** [comment]
I like that idea for MappingConfiguration + Profile

---
For `mapDefer`, I personally use `afterMap` for that. I'll put back `mapDefer` and see if the typings issue improve or not. Thanks all

**Comment by nartc** [comment]
@AliYusuf95 
![image](https://user-images.githubusercontent.com/25516557/158002958-f8bf13a6-1701-4414-8511-e55cdb5bffba.png)
![image](https://user-images.githubusercontent.com/25516557/158002959-7f6c3b27-b0f3-40fa-b707-73798fce4b25.png)
`typeConverter` with Array type is implemented

**Comment by micalevisk** [comment]
@nartc will you (+ community) keep maintaining v7 or just v8 after releasing it?

**Comment by nartc** [comment]
@micalevisk I'll still fix bugs for v7.

**Comment by nartc** [comment]
In v8, `pojos` will work THE SAME as `classes`.

I’m pretty happy with the changes so far (of course, the breaking changes will be the headache). The flow is so much clearer. Things (metadata discovery, applying metadata, which operation is important for which part…) happen in a predictable and precise way.

Performance will also be improving because metadata is applied once when `createMap` happens instead of on every single `.map()`

`core` becomes the main driving force. The strategies (aka ex-plugin) are ridiculously simple. All they do is: discover metadata, retrieve metadata, and applying metadata (with a default `applyMetadata` function provided by `core` that just works).

**Comment by nartc** [comment]
I'm looking for a tool to write automatic migrations (as well as acting as a CLI tool for `@automapper`), anyone have any suggestion? is `jscodeshift` a good tool?

**Comment by micalevisk** [comment]
I didn't dig much on this but I feel that `jscodeshift` is the way.

**Comment by nartc** [comment]
@micalevisk I dug into jscodeshift a bit yesterday and it maybe not the right choice. Looking into ts-morph today 😄

**Comment by AliYusuf95** [comment]
I don't have previous expertise in dealing with code migration using AST 😅

**Comment by AliYusuf95** [comment]
While it seems an interesting task but I think this may doesn't fit all projects.
In my case to migrate from v5 to v6 and v7 I wrote a code that allows me to continue to use the old usage while gradually migrating to the new usage at the same time I can use the new features 😄.

**Comment by nartc** [comment]
@AliYusuf95 I see. I’m a little hesitant to push for v8 because of the breaking changes. I’ll try to figure something out.

—-

Happy to report that v8 is 2-3 times faster than v7 with the following simple bench:
- 10000 objects mapped
- 20 iterations
- Avg on v7: ~110-120ms; avg on v8: ~41-43ms
- This was done in Angular (the web) so in NodeJS, it might be slightly faster.

Bundle-size of v8 is also smaller (12kb gzipped for core and classes) compared to 16.5kb gzipped of v7. V8’s core is inherently bigger than v7 but the functions are tree-shaken away.


I’m considering:
- Drop support for node 12
- Ship cjs and esm (no umd)
> These two points are already done on 8.0.0-beta.2; if you want commonjs output only, try 8.0.0-beta.0)

—-

#### convertUsing

I dropped the 2nd argument for convertUsing in v8. This is to select the source data for Converter. 

// TODO add example to demonstrate the change

#### Async

Things might be wonky with Async still. Haven’t found a good way to handle it

#### forSelf

Accept a Source or a Mapping.

// TODO add example

#### extend

Accept a pair of Models to extend or a Mapping to extend

#### createMap

createMap returns a Mapping (instead of FluentAPI) so it can be passed into extend()

// TODO add example

#### map*** operations

TBA

#### NestJS integration

TBA

#### Transformer Plugin

- Runs 40-60% faster than v7 due to removals of unnecessary operations on the AST

**Comment by micalevisk** [comment]
> Drop support for node 12

:+1: 

![image](https://user-images.githubusercontent.com/13461315/159176288-07adb452-0c31-4dd9-b2d3-82f08eafcd2e.png)

https://github.com/nodejs/Release

**Comment by nartc** [comment]
https://github.com/nartc/nx-post-mikro/commit/9564c617ee55d436a0f54abafdc9ba89fa5169ab
Check this commit out for a migration example.

**Comment by nartc** [comment]
![image](https://user-images.githubusercontent.com/25516557/159841207-57c2e0c4-b4da-4ee2-bee4-bbc9226dbcc5.png)
New docs + migration docs are in process.

## NestJS: `AutomapperModule.forRootAsync()`

Add support for `forRootAsync`, so that you can work with your ORMs. Here's an example with `MikroORM`

```ts
import { EntityManager } from '@mikro-orm/mongodb';

AutomapperModule.forRootAsync({
    inject: [EntityManager],
    useFactory: (em: EntityManager) => ({
        strategyInitializer: classes({
            // now you can use your ORM to determine how AutoMapper should construct the Destination
            destinationConstructor: (sourceObject, destinationIdentifier) =>
                em.create(destinationIdentifier, {}),
        }),
    }),
});
```

**Comment by nartc** [comment]
- `mapDefer` is back
- `undefinedSubstitution` is added cc @micalevisk

**Comment by nartc** [comment]
https://automapper-v8.netlify.app/docs/getting-started/migrate-to-automapper-v8 migration guide is up. cc @AliYusuf95 @micalevisk can you guys give it a quick proof read if you're available?

**Comment by nartc** [comment]
I've decided to release v8 a moment ago. Automatic Migration is nice to have but it's extremely cumbersome to write (took me 2 days already without much progress :( ). Instead, I tried to be very comprehensive in the Migration guide and hope that it is enough to get folks migrated. Most can be Find and Replace. Thank you all for all the discussions. v8 is great and I can't wait to hear from you all. Hopefully you all find the time (and find it worth it) to migrate to v8.
